control-plane/pkg/reconciler/consumergroup/consumergroup.go

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2021 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package consumergroup
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"knative.dev/pkg/reconciler"
+
+	kafkainternals "knative.dev/eventing-kafka-broker/control-plane/pkg/apis/internals/kafka/eventing/v1alpha1"
+	internalv1alpha1 "knative.dev/eventing-kafka-broker/control-plane/pkg/client/internals/kafka/clientset/versioned/typed/eventing/v1alpha1"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/client/internals/kafka/injection/reconciler/eventing/v1alpha1/consumergroup"
+	kafkainternalslisters "knative.dev/eventing-kafka-broker/control-plane/pkg/client/internals/kafka/listers/eventing/v1alpha1"
+)
+
+type Reconciler struct {
+	ConsumerLister  kafkainternalslisters.ConsumerLister
+	InternalsClient internalv1alpha1.InternalV1alpha1Interface
+}
+
+func (r Reconciler) ReconcileKind(ctx context.Context, cg *kafkainternals.ConsumerGroup) reconciler.Event {
+	if err := r.reconcileConsumers(ctx, cg); err != nil {
+		return err
+	}
+	cg.MarkReconcileConsumersSucceeded()
+
+	return nil
+}
+
+func (r Reconciler) reconcileConsumers(ctx context.Context, cg *kafkainternals.ConsumerGroup) error {
+
+	// Get consumers associated with the ConsumerGroup.
+	existingConsumers, err := r.ConsumerLister.Consumers(cg.GetNamespace()).List(labels.SelectorFromSet(cg.Spec.Selector))
+	if err != nil {
+		return cg.MarkReconcileConsumersFailed("ListConsumers", err)
+	}
+
+	// Stable sort consumers so that we give consumers different deletion
+	// priorities based on their state.
+	//
+	// Consumers at the tail of the list are deleted first.
+	sort.SliceStable(existingConsumers, func(i, j int) bool {
+		return existingConsumers[i].IsLessThan(existingConsumers[j])
+	})
+
+	nConsumers := int32(len(existingConsumers))
```

Comment:
Changes look good. 
Q: Will the scheduling of `cg.Spec.Replicas` and descheduling of `nConsumers-*cg.Spec.Replicas` be done in a separate PR?

Reply:
Yes, scheduling will be done in a separate PR